draft-ietf-quic-load-balancers.md

```diff
@@ -436,91 +436,180 @@ for decryption. To understand this algorithm, it is useful to define four
+Two functions are useful to define:
+
+The expand(length, pass, input_bytes) function concatenates three arguments and
+outputs 16 zero-padded octets. The first argument 'length' is an 8-bit integer
+that reports the sum of the configured nonce length and server id length, and
+forms the most significant octet of the output. It MUST NOT exceed 28. The
+second argument is an 8-bit integer that is the 'pass' of the algorithm, and
+forms the second-most significant octet of the output. The third argument is a
+variable-length stream of octets, which is copied into third-most significant
+octet of the output and beyond. The length of this octet stream is half the
+'length', rounded up. All remaining octets of the output are zero.
+
+For example,
 
 ~~~pseudocode
-expand_right(0xaaba3c, 0x0b, 0x02) =
-                  0x0b020000000000000000000000aaba3c
+expand(0x06, 0x02, 0xaaba3c) = 0x0602aaba3c0000000000000000000000
 ~~~
 
-Similarly, truncate_left() and truncate_right() take the most significant and
-least significant bits, respectively, from a ciphertext. For example, to take
-28 bits of a ciphertext:
+Similarly, truncate(input, n) returns the n most significant octets of 'input'.
 
 ~~~pseudocode
-truncate_left(0x2094842ca49256198c2deaa0ba53caa0, 28) = 0x2094842
-truncate_right(0x2094842ca49256198c2deaa0ba53caa0, 28) = 0xa53caa0
+truncate(0x2094842ca49256198c2deaa0ba53caa0, 28) = 0x2094842
 ~~~
 
+Let 'half_len' be equal to 'plaintext_len' / 2, rounded up.
+
+#### Algorithm Description
+
 The example at the end of this section helps to clarify the steps described
 below.
 
-1. The server concatenates the server ID and nonce to create plaintext_CID.
+1. The server concatenates the server ID and nonce to create plaintext_CID. The
+length of the result in octets is plaintext_len.
 
 2. The server splits plaintext_CID into components left_0 and right_0 of equal
-length, splitting an odd octet in half if necessary. For example,
-0x7040b81b55ccf3 would split into a left_0 of 0x7040b81 and right_0 of
-0xb55ccf3.
-
-3. Encrypt the result of expand_left(left_0, index)) to obtain a ciphertext,
-where 'index' is one octet: the two most significant bits of which are 0b00,
-and the six least significant bits are the length of the resulting connection
-ID in bytes, cid_len.
+length half_len. If plaintext_len is odd, right_0 clears its four most
+significant bits, and left_0 clears its four least significant bits. For
+example, 0x7040b81b55ccf3 would split into a left_0 of 0x7040b810 and right_0 of
+0x0b55ccf3.
+
+3. Encrypt the result of expand(plaintext_len, 1, left_0) using an AES-ECB-128
+cipher to obtain a ciphertext.
+
+4. XOR the most significant bits of the ciphertext with right_0 to form
+right_1. If plaintext_len is odd, clear the four most significant bits. Steps 3
+and 4 can be summarized as
+  
+```
+   right_1 = XOR(right_0,
+                truncate(AES_ECB(key, expand(plaintext_len, 1, left_0)),
+                         half_len)
```

Comment:
Assigning `truncate(AES_ECB(key, expand(plaintext_len, 1, left_0))` to an intermediate variable might make this fit line lengths easier.  If you don't, then this is poorly aligned with the line above.

This (and the later ones) appear to be missing a closing paren